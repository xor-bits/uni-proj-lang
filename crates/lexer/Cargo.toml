[package]
name = "lexer"
version.workspace = true
edition.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
# prob faster lexing + less boilerplate:
# logos = "0.13"

# only for unit tests
[dev-dependencies]
serde = { version = "1.0", features = ["derive"] }
insta = { version = "1.40", features = ["yaml"] }
