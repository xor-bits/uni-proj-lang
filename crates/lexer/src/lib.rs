#[cfg(test)]
use serde::{ser::SerializeStruct, Serialize};
use std::{fmt, ops::Range};

//

macro_rules! some {
    ($v:expr) => {
        if let Some(v) = $v {
            return Some(v);
        }
    };
}

macro_rules! exhaustive {
    ($($tok:ident::$var:ident $({ $($t:tt)* })?,)*) => {{
        use crate::Token::*;

        #[allow(unreachable_code)]
        fn _assert_exhaustive() {
            match unreachable!() {
                $($var { .. } => {})*
            }
        }

        [
            $($tok::$var $({ $($t)* })?),*
        ]
    }};
}

//

pub type Result<T, E = Error> = core::result::Result<T, E>;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Error {
    UnexpectedEoi,
    ExtraTokens,
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Error::UnexpectedEoi => f.write_str("unexpected end of input"),
            Error::ExtraTokens => f.write_str("found extra tokens"),
        }
    }
}

//

#[cfg_attr(test, derive(Serialize))]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Token {
    /// `;`
    Semi,

    /// `=`
    Assign,

    /// `(`
    LParen,

    /// `)`
    RParen,

    /// `{`
    LBrace,

    /// `}`
    RBrace,

    /// `+`
    Plus,

    /// `-`
    Minus,

    /// `*`
    Asterisk,

    /// `/`
    Slash,

    /// `%`
    Percent,

    /// `:`
    Colon,

    /// `,`
    Comma,

    /// `@`
    At,

    /// `&`
    Ampersand,

    /// `<`
    Lt,

    /// `>`
    Gt,

    /// `<=`
    Le,

    /// `>=`
    Ge,

    /// `==`
    Eq,

    /// `!=`
    Neq,

    /// `&&`
    And,

    /// `||`
    Or,

    /// `:=`
    Walrus,

    /// `->`
    RArrow,

    /// `fn`
    Fn,

    /// `for`
    For,

    /// `return`
    Return,

    /// `struct`
    Struct,

    /// `test`
    Test,

    /// single line comments like `// this is a comment`
    LineComment,

    /// identifiers like `_fn_name`
    Ident,

    /// floating point literal like `5.2`
    LitFloat,

    /// integer literal like `0xFFu8`
    LitInt,

    /// string literal like `"test"`
    LitStr,

    /// end of input
    Eoi,
}

impl Token {
    pub const fn enumerate() -> &'static [Self] {
        &exhaustive![
            Token::Semi,
            Token::Assign,
            Token::LParen,
            Token::RParen,
            Token::LBrace,
            Token::RBrace,
            Token::Plus,
            Token::Minus,
            Token::Asterisk,
            Token::Slash,
            Token::Percent,
            Token::Colon,
            Token::Comma,
            Token::At,
            Token::Ampersand,
            Token::Lt,
            Token::Gt,
            Token::Le,
            Token::Ge,
            Token::Eq,
            Token::Neq,
            Token::And,
            Token::Or,
            Token::Walrus,
            Token::RArrow,
            Token::Fn,
            Token::For,
            Token::Return,
            Token::Struct,
            Token::Test,
            Token::LineComment,
            Token::Ident,
            Token::LitFloat,
            Token::LitInt,
            Token::LitStr,
            Token::Eoi,
        ]
    }

    pub const fn token_type(self) -> TokenType {
        match self {
            Token::Semi => TokenType::Symbols(";"),
            Token::Assign => TokenType::Symbols("="),
            Token::LParen => TokenType::Symbols("("),
            Token::RParen => TokenType::Symbols(")"),
            Token::LBrace => TokenType::Symbols("{"),
            Token::RBrace => TokenType::Symbols("}"),
            Token::Plus => TokenType::Symbols("+"),
            Token::Minus => TokenType::Symbols("-"),
            Token::Asterisk => TokenType::Symbols("*"),
            Token::Slash => TokenType::Symbols("/"),
            Token::Percent => TokenType::Symbols("%"),
            Token::Colon => TokenType::Symbols(":"),
            Token::Comma => TokenType::Symbols(","),
            Token::At => TokenType::Symbols("@"),
            Token::Ampersand => TokenType::Symbols("&"),
            Token::Lt => TokenType::Symbols("<"),
            Token::Gt => TokenType::Symbols(">"),
            Token::Le => TokenType::Symbols("<="),
            Token::Ge => TokenType::Symbols(">="),
            Token::Eq => TokenType::Symbols("=="),
            Token::Neq => TokenType::Symbols("!="),
            Token::And => TokenType::Symbols("&&"),
            Token::Or => TokenType::Symbols("||"),
            Token::Walrus => TokenType::Symbols(":="),
            Token::RArrow => TokenType::Symbols("->"),
            Token::Fn => TokenType::Keyword("fn"),
            Token::For => TokenType::Keyword("for"),
            Token::Return => TokenType::Keyword("return"),
            Token::Struct => TokenType::Keyword("struct"),
            Token::Test => TokenType::Keyword("test"),
            Token::LineComment => TokenType::Other,
            Token::Ident => TokenType::Other,
            Token::LitFloat => TokenType::Other,
            Token::LitInt => TokenType::Other,
            Token::LitStr => TokenType::Other,
            Token::Eoi => TokenType::Other,
        }
    }
}

impl fmt::Display for Token {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self.token_type() {
            TokenType::Symbols(s) => f.write_str(s),
            TokenType::Keyword(s) => f.write_str(s),
            TokenType::Other => write!(f, "{self:?}"),
        }
    }
}

//

#[derive(Debug, Clone, Copy)]
pub enum TokenType {
    Symbols(&'static str),
    Keyword(&'static str),
    Other,
}

//

#[cfg_attr(test, derive(Serialize))]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct SpannedToken<'a> {
    token: Token,
    span: Span<'a>,
}

impl<'a> SpannedToken<'a> {
    pub const fn from_whole_str(token: Token, str: &'a str) -> Self {
        Self {
            token,
            span: Span::from_whole_str(str),
        }
    }

    pub fn token(&self) -> Token {
        self.token
    }

    pub fn span(&self) -> Range<usize> {
        self.span.span()
    }

    pub fn as_str(&self) -> &'a str {
        self.span.as_str()
    }
}

impl fmt::Display for SpannedToken<'_> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        if self.token == Token::Eoi {
            write!(f, "<eoi>")
        } else {
            f.write_str(self.as_str())
        }
    }
}

//

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Span<'a> {
    from: usize,
    to: usize,
    source: &'a str,
}

impl<'a> Span<'a> {
    pub const fn from_whole_str(str: &'a str) -> Self {
        Self {
            from: 0,
            to: str.len(),
            source: str,
        }
    }

    pub fn span(&self) -> Range<usize> {
        self.from..self.to
    }

    pub fn as_str(&self) -> &'a str {
        &self.source[self.span()]
    }
}

#[cfg(test)]
impl Serialize for Span<'_> {
    fn serialize<S>(&self, serializer: S) -> core::result::Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut s = serializer.serialize_struct("SpannedToken", 3)?;
        s.serialize_field("range", &format!("{:?}", self.span()))?;
        s.serialize_field("str", &self.as_str())?;
        s.end()
    }
}

//

#[derive(Debug, Clone)]
pub struct Lexer<'a> {
    source: &'a str,
    at: &'a str,
    eoi: bool,
    err: bool,
}

impl<'a> Lexer<'a> {
    pub const fn new(source: &'a str) -> Self {
        Self {
            source,
            at: source,
            eoi: false,
            err: false,
        }
    }

    // pub const fn error(&self) -> Result<()> {
    //     self.err
    // }

    fn src_index(&self) -> usize {
        self.at.as_ptr() as usize - self.source.as_ptr() as usize
    }

    fn advance_by(&mut self, len: usize) {
        self.at = &self.at[len..];
    }

    fn advance(&mut self, span: &Span) {
        self.advance_by(span.span().len())
    }

    fn span_to(&self, len: usize) -> Span<'a> {
        let at = self.src_index();
        Span {
            from: at,
            to: at + len,
            source: self.source,
        }
    }

    fn spanned_token_to(&mut self, token: Token, len: usize) -> Option<Result<SpannedToken<'a>>> {
        let span = self.span_to(len);
        self.advance(&span);
        Some(Ok(SpannedToken { token, span }))
    }

    fn try_match(&mut self, token: Token) -> Option<Result<SpannedToken<'a>>> {
        let TokenType::Symbols(str) = token.token_type() else {
            return None;
        };
        if !self.at.starts_with(str) {
            return None;
        }

        self.spanned_token_to(token, str.len())
    }

    fn try_match_number(&mut self) -> Option<Result<SpannedToken<'a>>> {
        let first = self.at.chars().next()?;

        if !first.is_ascii_digit() {
            return None;
        }

        let mut dot_found = false;
        let term = self.at[1..]
            .find(|c: char| {
                !c.is_ascii_digit()
                    || if dot_found {
                        false
                    } else if c == '.' {
                        dot_found = true;
                        true
                    } else {
                        false
                    }
            })
            .map(|term| term + 1)
            .unwrap_or(self.at.len());

        let token = if dot_found {
            Token::LitFloat
        } else {
            Token::LitInt
            // let num = self.at[..term].parse::<i128>().unwrap(); // the unwrap should never fail
            // Token::LitInt(num)
        };

        self.spanned_token_to(token, term)
    }

    fn try_match_str(&mut self) -> Option<Result<SpannedToken<'a>>> {
        if !self.at.starts_with('"') {
            return None;
        }

        let Some(term) = self.at[1..].find('"') else {
            return self.err(Error::UnexpectedEoi);
        };

        self.spanned_token_to(Token::LitStr, term + 2)
    }

    fn try_match_ident(&mut self) -> Option<Result<SpannedToken<'a>>> {
        let first = self.at.chars().next()?;

        if !(first == '_' || first.is_alphabetic()) {
            return None;
        }

        let term = self.at[1..]
            .find(|c: char| !(c.is_alphanumeric() || c == '_'))
            .map(|term| term + 1)
            .unwrap_or(self.at.len());

        self.spanned_token_to(Token::Ident, term)
    }

    fn try_match_comment(&mut self) -> Option<Result<SpannedToken<'a>>> {
        if !self.at.starts_with("//") {
            return None;
        }

        let term = self.at[2..]
            .lines()
            .next()
            .map(|line| line.len() + 2)
            .unwrap_or(self.at.len());

        self.spanned_token_to(Token::LineComment, term)
    }

    fn err(&mut self, err: Error) -> Option<Result<SpannedToken<'a>>> {
        self.eoi = true;
        Some(Err(err))
    }

    fn eoi(&mut self) -> Option<Result<SpannedToken<'a>>> {
        self.eoi = true;
        self.spanned_token_to(Token::Eoi, 0)
    }
}

impl<'a> Iterator for Lexer<'a> {
    type Item = Result<SpannedToken<'a>>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.eoi {
            return None;
        }
        if self.err {
            return self.eoi();
        }
        self.at = self.at.trim();

        // tokens in the order of 'more specific -> less specific' to emulate priority
        // for example := should be parsed as Init instead of Type and Assign

        // custom tokens
        some!(self.try_match_str());
        some!(self.try_match_comment());
        some!(self.try_match_number());

        // simple const str tokens
        for (_, token) in Token::enumerate()
            .iter()
            .rev()
            .copied()
            .filter_map(|tok| match tok.token_type() {
                TokenType::Symbols(s) => Some((tok, s)),
                _ => None,
            })
            // just a debug assert to make sure that the tokens are in the correct order
            .scan(usize::MAX, |prev, (now, s)| {
                debug_assert!(*prev >= s.len());
                Some((s.len(), now))
            })
        {
            some!(self.try_match(token));
        }

        // idents after keywords
        if let Some(ident) = self.try_match_ident() {
            let ident = match ident {
                Ok(ident) => ident,
                Err(err) => return Some(Err(err)),
            };

            for (s, token) in
                Token::enumerate()
                    .iter()
                    .rev()
                    .copied()
                    .filter_map(|tok| match tok.token_type() {
                        TokenType::Keyword(s) => Some((s, tok)),
                        _ => None,
                    })
            {
                if ident.as_str() == s {
                    return Some(Ok(SpannedToken {
                        token,
                        span: ident.span,
                    }));
                }
            }

            return Some(Ok(ident));
        }

        if !self.at.is_empty() {
            return self.err(Error::ExtraTokens);
        }

        self.eoi()
    }
}

//

#[cfg(test)]
mod tests {
    use crate::{Lexer, Result, SpannedToken, Token, TokenType};

    use insta::assert_yaml_snapshot;
    use serde::Serialize;

    fn lex(str: &str) -> impl Serialize + '_ {
        let tokens: Result<Vec<SpannedToken>> = Lexer::new(str).collect();
        (str, tokens.map_err(|err| err.to_string()))
    }

    #[test]
    fn lex_test() {
        assert_yaml_snapshot!(lex(
            "in->\"vali\"@->:d synt(ax} {but) t;h:e l=e:=xer should handle it"
        ));
    }

    #[test]
    fn lex_str_lit() {
        assert_yaml_snapshot!(lex("_a2"));
        assert_yaml_snapshot!(lex(""));
        assert_yaml_snapshot!(lex("\""));
        assert_yaml_snapshot!(lex("\"a"));
        assert_yaml_snapshot!(lex("\"a\""));
    }

    #[test]
    fn lex_all_simple() {
        let all_tokens: String = Token::enumerate()
            .iter()
            .filter_map(|tok| match tok.token_type() {
                TokenType::Symbols(s) | TokenType::Keyword(s) => Some(format!("{s} ")),
                TokenType::Other => None,
            })
            .collect();

        assert_yaml_snapshot!(lex(&all_tokens));
    }

    #[test]
    fn lex_all_no_spaces() {
        let all_tokens: String = Token::enumerate()
            .iter()
            .filter_map(|tok| match tok.token_type() {
                TokenType::Symbols(s) | TokenType::Keyword(s) => Some(s),
                TokenType::Other => None,
            })
            .collect();

        assert_yaml_snapshot!(lex(&all_tokens));
    }

    #[test]
    fn lex_comment() {
        assert_yaml_snapshot!(lex("// comment\ncode"));
    }

    #[test]
    fn lex_example() {
        assert_yaml_snapshot!(lex(include_str!("../../../tests/trivial")));
    }
}
