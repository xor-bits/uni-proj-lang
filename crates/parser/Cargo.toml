[package]
name = "parser"
version.workspace = true
edition.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
lexer.path = "../lexer"
macros.path = "../macros"

# only for unit tests
[dev-dependencies]
serde = { version = "1.0", features = ["derive"] }
insta = { version = "1.34", features = ["yaml"] }
