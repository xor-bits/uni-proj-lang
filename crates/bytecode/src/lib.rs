#![feature(inline_const_pat)]

//

use std::{thread, time::Duration};

use lexer::Lexer;
use parser::{
    ast::{Ast, Root},
    ParseStream, Result,
};

#[cfg(test)]
use serde::Serialize;

//

pub struct Runtime {
    code: Vec<u8>,
    strings: Vec<u8>,

    fuel: usize,
    fuel_enabled: bool,

    ip: usize,
    stack: Vec<u8>,
    regs: Vec<Value>,
}

impl Runtime {
    pub const fn new() -> Self {
        Self {
            code: Vec::new(),
            strings: Vec::new(),

            fuel: 0,
            fuel_enabled: false,

            ip: 0,
            stack: Vec::new(),
            regs: Vec::new(),
        }
    }

    pub fn add_fuel(&mut self, added_fuel: usize) {
        self.fuel = self.fuel.saturating_add(added_fuel);
    }

    pub fn enable_fuel(&mut self) {
        self.fuel_enabled = true;
    }

    pub fn disable_fuel(&mut self) {
        self.fuel_enabled = false;
    }

    pub fn fuel_enabled(&self) -> bool {
        self.fuel_enabled
    }

    pub fn assemble(&mut self, instr: &[Instruction]) -> Result<()> {
        if self.code.is_empty() {
            self.code.extend_from_slice(&[0xDE, 0xAF, 0xC0, 0xDE]);
        }

        self.strings.extend_from_slice(&[0, 1]);
        self.strings.extend_from_slice(&[1, 13]);
        self.strings.extend_from_slice(b" ");
        self.strings.extend_from_slice(b"Hello, world!");

        for ins in instr {
            ins.as_bytes(&mut self.code);
        }

        Ok(())
    }

    pub fn run(&mut self, code: &str) -> Result<()> {
        let mut parser = ParseStream::from_lexer(Lexer::new(code));
        let ast: Ast<Root> = parser.parse()?;

        if self.code.is_empty() {
            self.code.extend_from_slice(&[0xDE, 0xAF, 0xC0, 0xDE]);
        }

        self.strings.extend_from_slice(&[0, 1]);
        self.strings.extend_from_slice(&[1, 13]);
        self.strings.extend_from_slice(b" ");
        self.strings.extend_from_slice(b"Hello, world!");

        for ins in [
            // set X = 0 and Y = 1
            Instruction::I32Const { val: 0 },
            Instruction::I32Pop { reg: 0 },
            Instruction::I32Const { val: 1 },
            Instruction::I32Pop { reg: 1 },
            // print X
            Instruction::I32Push { reg: 0 },
            Instruction::I32Print,
            // tmp = X + Y
            Instruction::I32Push { reg: 0 },
            Instruction::I32Push { reg: 1 },
            Instruction::I32Add,
            // X = Y and Y = tmp
            Instruction::I32Push { reg: 1 },
            Instruction::I32Pop { reg: 0 },
            Instruction::I32Pop { reg: 1 },
            // loop
            Instruction::Step,
            Instruction::GotoRelI8 { offs: -15 },
        ] {
            ins.as_bytes(&mut self.code);
        }

        println!("{:?}", self.code);

        self.exec();

        Ok(())
    }

    pub fn reset(&mut self) {
        self.ip = 0usize;
        self.stack.clear();
        self.regs.clear();
    }

    pub fn stack(&self) -> &[u8] {
        &self.stack[..]
    }

    pub fn registers(&self) -> &[Value] {
        &self.regs[..]
    }

    pub fn exec(&mut self) {
        let [0xDE, 0xAF, 0xC0, 0xDE, code @ ..] = self.code.as_slice() else {
            panic!("invalid bytecode");
        };

        loop {
            if self.fuel_enabled {
                if self.fuel == 0 {
                    panic!("out of fuel");
                }
                self.fuel -= 1;
            }

            if self.ip >= code.len() {
                panic!("ip out of bounds");
            }

            // println!("ip = {ip} opcode = {}", code[ip]);
            let ins = Instruction::from_bytes(&code[self.ip..]).expect("invalid opcode");
            // println!("instr = {ins:?}");

            match ins {
                Instruction::I32Const { val: small_const } => self
                    .stack
                    .extend_from_slice(&(small_const as i32).to_ne_bytes()[..]),
                Instruction::I32Load8 { reg } => {}
                Instruction::I32Push { reg } => {
                    let Value::I32(val) =
                        *self.regs.get(reg as usize).expect("register out of bounds")
                    else {
                        panic!("invalid register cast");
                    };

                    Self::push_bytes(&mut self.stack, val.to_ne_bytes());
                }
                Instruction::I32Pop { reg } => {
                    let val = Self::pop_bytes(&mut self.stack);
                    let val = Value::I32(i32::from_ne_bytes(val));

                    if let Some(reg) = self.regs.get_mut(reg as usize) {
                        *reg = val;
                    } else {
                        self.regs.resize_with(reg as usize, || Value::Unknown);
                        self.regs.push(val);
                    }
                }
                Instruction::I32Add => {
                    let a = i32::from_ne_bytes(Self::pop_bytes(&mut self.stack));
                    let b = i32::from_ne_bytes(Self::pop_bytes(&mut self.stack));
                    let res = a.wrapping_add(b);

                    Self::push_bytes(&mut self.stack, i32::to_ne_bytes(res));
                }
                Instruction::I32Print => {
                    let i = i32::from_ne_bytes(Self::pop_bytes(&mut self.stack));
                    println!("{i}");
                }
                Instruction::I32Dup => {
                    let val: [u8; 4] = Self::pop_bytes(&mut self.stack);
                    Self::push_bytes(&mut self.stack, val);
                    Self::push_bytes(&mut self.stack, val);
                }
                Instruction::GotoRelI8 { offs } => {
                    self.ip = self
                        .ip
                        .checked_add_signed(offs as isize)
                        .expect("ip out of bounds");
                    continue;
                }
                Instruction::Debug => {
                    println!("ip={:#x} {:?} {:?}", self.ip, self.stack, self.regs);
                }
                Instruction::Step => {
                    self.ip += ins.size();
                    return;
                }
            }

            self.ip += ins.size();
        }
    }

    fn pop_bytes<const N: usize>(stack: &mut Vec<u8>) -> [u8; N] {
        let res = *stack.last_chunk().expect("stack underflow");
        stack.truncate(stack.len() - N);
        res
    }

    fn push_bytes<const N: usize>(stack: &mut Vec<u8>, bytes: [u8; N]) {
        stack.extend_from_slice(&bytes[..]);
    }
}

impl Default for Runtime {
    fn default() -> Self {
        Self::new()
    }
}

//

#[cfg_attr(test, derive(Serialize))]
#[derive(Debug, Clone, Copy)]
pub enum Value {
    I32(i32),
    Unknown,
}

//

macro_rules! impl_instructions {
    (
        $(#[$($attrs:tt)*])*
        pub enum $id:ident {
            $(
                $(#[$($variant_attrs:tt)*])*
                $variant_id:ident $( {$($field_id:ident : $field_type:ty,)*} )?,
            )*
        }
    ) => {
        $(#[$($attrs)*])*
        pub enum $id {
            $(
                $(#[$($variant_attrs)*])*
                $variant_id $( {$($field_id : $field_type),*} )?,
            )*
        }

        #[derive(Debug, Clone, Copy)]
        pub enum OpCode {
            $(
                $(#[$($variant_attrs)*])*
                $variant_id,
            )*
        }

        impl $id {
            pub fn from_bytes(bytes: &[u8]) -> Option<Self> {
                Some(match bytes {
                    $(
                        [const { OpCode::$variant_id as usize as u8 }, _ops @ ..] => {
                            $(
                                let mut _ops: &[u8] = _ops;
                                $(
                                    let field;
                                    (field, _ops) = _ops.split_first_chunk().unwrap();
                                    let $field_id = <$field_type>::from_le_bytes(*field);
                                )*
                            )?
                            $id::$variant_id $( { $($field_id )* } )?
                        }
                    )*
                    _ => return None,
                })
            }

            pub const fn all() -> impl IntoIterator<Item = Self> {
                [
                    $(
                        $id::$variant_id $({$($field_id: 0,)*})?,
                    )*
                ]
            }

            pub const fn size(self) -> usize {
                match self {
                    $(
                        $id::$variant_id { .. } => 1 $($(+ std::mem::size_of::<$field_type>())*)?,
                    )*
                }
            }

            pub fn as_bytes(self, dst: &mut Vec<u8>) {
                match self {
                    $(
                        $id::$variant_id $( { $($field_id)* } )? => {
                            dst.reserve(self.size());
                            dst.push(OpCode::$variant_id as usize as u8);
                            $($(
                                dst.extend_from_slice(&$field_id.to_le_bytes()[..]);
                            )*)?
                        },
                    )*
                }
            }

            pub const fn as_opcode(self) -> OpCode {
                match self {
                    $(
                        Self::$variant_id { .. } => OpCode::$variant_id,
                    )*
                }
            }
        }
    };
}

impl_instructions! {

#[derive(Debug, Clone, Copy)]
// #[repr(u8)]
pub enum Instruction {
    /// push a constant to stack
    I32Const {
        val: i8,
    },

    /// push a constant to stack from the i32 const pool
    I32Load8 {
        reg: u8,
    },

    /// push a local to stack
    I32Push {
        reg: u8,
    },

    /// pop stack to a local
    I32Pop {
        reg: u8,
    },

    /// pop two i32's and push the sum
    I32Add,

    /// pop a i32 and print it
    I32Print,

    /// duplicate the top of stack (i32)
    I32Dup,

    /// goto relative address
    GotoRelI8 {
        offs: i8,
    },

    /// debug the program
    Debug,

    /// step the program
    Step,
}

}

//

#[cfg(test)]
mod tests {
    use insta::assert_yaml_snapshot;

    use crate::{Instruction, Runtime};

    #[test]
    fn fibonacci_series() {
        let instr = [
            // set X = 0 and Y = 1
            Instruction::I32Const { val: 0 },
            Instruction::I32Pop { reg: 0 },
            Instruction::I32Const { val: 1 },
            Instruction::I32Pop { reg: 1 },
            // print X
            Instruction::Step,
            // Instruction::I32Push { reg: 0 },
            // Instruction::I32Print,
            // tmp = X + Y
            Instruction::I32Push { reg: 0 },
            Instruction::I32Push { reg: 1 },
            Instruction::I32Add,
            // X = Y and Y = tmp
            Instruction::I32Push { reg: 1 },
            Instruction::I32Pop { reg: 0 },
            Instruction::I32Pop { reg: 1 },
            // loop
            Instruction::GotoRelI8 { offs: -12 }, // -15
        ];

        let mut runtime = Runtime::new();
        runtime.assemble(&instr[..]).unwrap();

        let mut results = Vec::new();

        for _ in 0..16 {
            runtime.exec();
            results.push(runtime.registers()[0]);
        }

        assert_yaml_snapshot!(results);
    }
}
