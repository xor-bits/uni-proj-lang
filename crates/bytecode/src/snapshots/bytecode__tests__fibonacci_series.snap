---
source: crates/bytecode/src/lib.rs
expression: results
---
- I32: 0
- I32: 1
- I32: 1
- I32: 2
- I32: 3
- I32: 5
- I32: 8
- I32: 13
- I32: 21
- I32: 34
- I32: 55
- I32: 89
- I32: 144
- I32: 233
- I32: 377
- I32: 610
